packages/tupaia-web/src/api/mutations/useLogin.ts

```diff
@@ -0,0 +1,35 @@
+/*
+ * Tupaia
+ *  Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { useNavigate } from 'react-router-dom';
+import { useMutation, useQueryClient } from 'react-query';
+import { post } from '../api';
+
+type LoginCredentials = {
+  email: string;
+  password: string;
+};
+export const useLogin = () => {
+  const navigate = useNavigate();
+  const queryClient = useQueryClient();
+
+  return useMutation<any, Error, LoginCredentials, unknown>(
+    ({ email, password }: LoginCredentials) => {
+      return post('login', {
+        data: {
+          emailAddress: email,
+          password,
+          deviceName: window.navigator.userAgent,
+        },
+      });
+    },
+    {
+      onSuccess: () => {
+        queryClient.invalidateQueries();
+        navigate(-1);
```

Comment:
So I think technically, if someone went straight to `/login` and then it navigated back to `-1`, it means that the user would be taken to whether they were before coming to Tupaia. So potentially we might need to check whether there is another route in the history, and if not, go to the default project?

Reply:
Very good point. Thanks for thinking about that. I have add a util for safely navigating back and used it in the auth modal for closing and also in the login on success.